GetEnvironmentVariable: Avoid StringBuilder marshaling overhead

justinvp opened the pull request:
Try a stack allocated buffer first, then fallback to using the shared array pool.

Results on my machine:


|             Method |      Mean |     Error |    StdDev |  Gen 0 | Allocated |
|------------------- |----------:|----------:|----------:|-------:|----------:|
|  `"TMP"` Before | 156.51 ns | 2.0132 ns | 1.8831 ns | 0.0875 |     368 B |
|  `"TMP"` After |  92.88 ns | 0.5002 ns | 0.4434 ns | 0.0209 |      88 B |
| `"PATH"` Before | 621.59 ns | 2.2668 ns | 2.1204 ns | 0.9565 |    4016 B |
| `"PATH"` After | 287.02 ns | 1.0377 ns | 0.8666 ns | 0.2704 |    1136 B |


`"TMP"` is under the stack limit and `"PATH"` is over the limit (uses the pool).

justinvp commented:
@dotnet-bot test Windows_NT x86 full_opt legacy_backend CoreCLR Perf Tests Correctness please

justinvp commented:
@dotnet-bot test this please

danmosemsft commented on src/mscorlib/src/System/Environment.cs:
```diff
@@ -536,23 +537,33 @@ private static void ValidateTarget(EnvironmentVariableTarget target)
         {
-            StringBuilder sb = StringBuilderCache.Acquire(128); // A somewhat reasonable default size
-            int requiredSize = Win32Native.GetEnvironmentVariable(variable, sb, sb.Capacity);
+            Span<char> buffer = stackalloc char[128]; // A somewhat reasonable default size
```
is it reasonable? since this is now a stack allocation, maybe it's cheap to be more generous? although, this is probably fine for most variables except path, include, libpath, etc.

danmosemsft replied:
@stephentoub do we have a rule of thumb for reasonable stackalloc max? In the CoreFX code I see mostly a few bytes, a cutoff of 100, 200, one of 256 is the largest.

justinvp replied:
I had the same question. 128 chars (256 bytes) seemed like a reasonable limit... but I could see going a little higher. Not sure how far we want to push it, though.

stephentoub replied:
We don't have a hard and fast rule, but we've generally stayed at <= 256 bytes.

danmosemsft commented on src/mscorlib/src/System/Environment.cs:
```diff
@@ -536,23 +537,33 @@ private static void ValidateTarget(EnvironmentVariableTarget target)
+            int requiredSize = Win32Native.GetEnvironmentVariable(variable, buffer);
 
-            if (requiredSize == 0 && Marshal.GetLastWin32Error() == Win32Native.ERROR_ENVVAR_NOT_FOUND)
+            if (requiredSize > buffer.Length)
```
confirmed that GetEnvironmentVariable returns the length including terminating null. StringBuilder.Capacity does not include terminating null so that would not have mattered before.

danmosemsft commented on src/mscorlib/src/Microsoft/Win32/Win32Native.cs:
```diff
@@ -687,7 +687,15 @@ internal class WIN32_FIND_DATA
 
         [DllImport(Interop.Libraries.Kernel32, CharSet = CharSet.Auto, SetLastError = true, BestFitMapping = false)]
-        internal static extern int GetEnvironmentVariable(string lpName, [Out]StringBuilder lpValue, int size);
+        private static extern unsafe int GetEnvironmentVariable(string lpName, char* lpValue, int size);
```
you've also eliminated the StringBuilder marshaling it was doing before, which forced a copy.
https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/interop-pinvokes.md

justinvp replied:
Yup :)

danmosemsft commented:
@dotnet-bot test Windows_NT x86 full_opt legacy_backend CoreCLR Perf Tests Correctness please (nolog)

jkotas commented on src/mscorlib/src/System/Environment.cs:
```diff
@@ -536,23 +537,33 @@ private static void ValidateTarget(EnvironmentVariableTarget target)
-                sb.Capacity = requiredSize;
-                sb.Length = 0;
-                requiredSize = Win32Native.GetEnvironmentVariable(variable, sb, sb.Capacity);
+                return Marshal.GetLastWin32Error() == Win32Native.ERROR_ENVVAR_NOT_FOUND ? null : string.Empty;
```
Nit: I would move the check for error condition right after the PInvoke to make it easier to see that there is nothing modifying `Marshal.GetLastWin32Error()`.

justinvp replied:
Done. I also changed the condition back to the way it was as the string ctor will take care of returning `string.Empty` for empty spans (when `requiredSize == 0` and the error isn’t “not found”).

justinvp commented:
I had added `clearArray: true` when returning the array to the pool out of caution and it didn't appear to significantly impact the perf, but re-running the microbenchmarks again does show a slowdown.

`clearArray: true`:

|                Method |      Mean |     Error |    StdDev |  Gen 0 | Allocated |
|---------------------- |----------:|----------:|----------:|-------:|----------:|
|    Before | 702.23 ns | 2.0626 ns | 1.8284 ns | 1.1282 |    4736 B |
|    After | 723.63 ns | 2.8812 ns | 2.5541 ns | 0.3271 |    1376 B |

`clearArray: false`:

|                Method |      Mean |      Error |     StdDev |  Gen 0 | Allocated |
|---------------------- |----------:|-----------:|-----------:|-------:|----------:|
|    Before | 684.89 ns |  3.7051 ns |  3.2845 ns | 1.1282 |    4736 B |
|    After | 316.44 ns |  1.9137 ns |  1.7901 ns | 0.3276 |    1376 B |

Thus, I've added a commit that removes `clearArray: true` since:

 - Any code in the process can already read environment variables whenever it wants
 - Clearing the array before returning it to the pool is not done anywhere else in corelib
 - CoreRT's `ExpandEnvironmentVariables` implementation uses the pool and [doesn't clear the array](https://github.com/dotnet/corert/blob/6a9700085acec28d32d14ab89bcf9cd91ff1e581/src/System.Private.CoreLib/src/System/Environment.Windows.cs#L46)

Let me know if you have any concerns
